buildenv/jenkins/common/build.groovy

```diff
@@ -326,8 +326,64 @@ def archive_sdk() {
     }
 }
 
+/*
+* When crash occurs in java, there will be an output in the log: IEATDUMP success for DSN='JENKINS.JVM.JENKIN*.*.*.X&DS'.
+* The X&DS is a macro that is passed to the IEATDUMP service on z/OS.
+* The dump is split into multiple files by the macro if the address space is large.
+* Each dataset is named in a similar way but there is a distinct suffix to specify the order. 
+* For example, JVM.*.*.*.001, JVM.*.*.*.002, and JVM.*.*.*.003.
+* The function fetchFile moves each dataset into the current directory of the machine, and appends the data to the core.xxx.dmp at the same time.
+* The file core.xxx.dmp is uploaded to artifactory for triage.
+*/
+def fetchFile(src, dest) {
+    // remove the user name from the dataset name
+    def offset = USER.length() + 1
+    def fileToMove = src.substring(src.indexOf("${USER}.") + offset)
+    // move a dataset to the current directory
+    sh "mv '//${fileToMove}' ${src}"
+    // append the dataset to core.xxx.dmp file
+    sh "cat ${src} >> ${dest}"
+    // remove the dataset from the current directory
+    sh "rm ${src}"
+    
+}
+
 def archive_diagnostics() {
-    sh "find . -name 'core.*.dmp' -o -name 'javacore.*.txt' -o -name 'Snap.*.trc' -o -name 'jitdump.*.dmp' | tar -zcvf ${DIAGNOSTICS_FILENAME} -T -"
+    if (SPEC.contains('zos')) {
+        def logContent = currentBuild.rawBuild.getLog()
+        // search for each occurrence of IEATDUMP success for DSN=
+        def matches = logContent =~ /IEATDUMP success for DSN=.*/
+        for (match in matches) {
+            // extract the naming convention of datasets
+            def beginIndex = match.indexOf('\'') + 1
+            def endIndex = match.lastIndexOf('.')
+            def filename = match.substring(beginIndex, endIndex)
+            // count how many datasets created by the macro
+            def num = sh returnStdout: true, script: "tso listcat | grep ${filename} | wc -l"
+            def numFiles = num.trim()
+            numFiles = numFiles.toInteger()
+            numFiles ++
+            for (i = 1; i < numFiles; i++) {
```

Comment:
It might be clearer to remove the increment of `numFiles` and adjust the loop condition to `i <= numFiles`.